tensorboard/plugins/profile/pod_viewer/topology_graph/topology-graph.ts

```diff
@@ -0,0 +1,573 @@
+   */
+  getChipYLoc: function(yHost: number, yChip: number): number {
+    return yHost * this._hostGridHeight + HOST_TO_CHIP_MARGIN +
+        yChip * (CHIP_GRID_SIZE + CHIP_TO_CHIP_MARGIN);
+  },
+  /**
+   * Returns the x-axis location for the xNode'th node of the xChip'th chip of
+   * the xHost'th host.
+   */
+  getNodeXLoc: function(xHost: number, xChip: number, xNode: number): number {
+    return this.getChipXLoc(xHost, xChip) + xNode * this._nodeGridWidth;
+  },
+  /**
+   * Returns the location for each host in the system.
+   */
+  createHostData: function(
+      hostXDim: number, hostYDim: number): Array<HostData> {
+    let hostData = [];
+    for (let i = 0; i < hostXDim; i++) {
+      for (let j = 0; j < hostYDim; j++) {
+        hostData.push({xdim: i, ydim: j});
+      }
+    }
+    return hostData;
+  },
+  /**
+   * Draw the labels on x-axis and y-axis.
+   */
+  drawLabels: function(svg: any, xdims: number[], ydims: number[]) {
+    // Draw label on x axis.
+    svg.selectAll('.xLabel').data(xdims).enter().append('text').text((d) => d)
+        .attr('x', (d, i) => this.getChipXLoc(
+                       Math.floor(i / this._hostXStride),
+                           i % this._hostXStride))
+        .attr('y', 0)
+        .style('text-anchor', 'middle')
+        .attr('transform', 'translate(' + CHIP_GRID_SIZE / 2 + ', -6)')
+        .attr('class', 'axis');
+
+    // Draw label on y axis.
+    svg.selectAll('.yLabel').data(ydims).enter().append('text').text((d) => d)
+        .attr('x', 0)
+        .attr('y', (d, i) => this.getChipYLoc(
+                       Math.floor(i / HOST_Y_STRIDE), i % HOST_Y_STRIDE))
+        .style('text-anchor', 'middle')
+        .attr('transform', 'translate(-12,' + CHIP_GRID_SIZE / 2 + ')')
+        .attr('class', 'axis');
+  },
+  /**
+   * Draw the UI of host cards.
+   */
+  drawHostCards: function(svg, data, gridWidth: number, gridHeight: number) {
+    let cards = svg.selectAll('.xdim').data(data, (d) => d.xdim);
+    cards.enter().append('rect')
+        .attr('x', (d) => d.xdim * gridWidth)
+        .attr('y', (d) => d.ydim * gridHeight)
+        .attr('rx', 4 * gridWidth / gridHeight)
+        .attr('ry', 4)
+        .attr('class', 'hour bordered')
+        .attr('width', gridWidth - HOST_TO_HOST_MARGIN)
+        .attr('height', gridHeight - HOST_TO_HOST_MARGIN)
+        .attr('border', 1)
+        .style('fill', 'F0F0F0')
+        .style('stroke', 'black')
+        .style('stroke-width', 1)
+        .merge(cards)
+        .transition()
+        .duration(1000);
+    cards.exit().remove();
+  },
+  /**
+   * Draw the UI of node cards.
+   */
+  drawNodeCards: function(svg: any, data: Array<TopoData>, colorScale: any) {
+    let cards = svg.selectAll('.xdim').data(data, (d) => d.xdim);
+    let parent = this;
+    cards.enter().append('rect')
+        .attr('id', (d) => 'rid' + d.rid)
+        .attr('x', (d) => {
+          return this.getNodeXLoc(
```

Comment:
re previous conversation about `data` change: so TopoData cannot change in a fashion that changes any of these values?

```ts
const data0 = [{xdim: 0, ydim: 3}, {xdim: 1, ydim: 4}];
const meow = svg.selectAll('.meow').data(data0, (d) => d.xdim);
meow.enter()
   ...//more code here

// Later
const data1 = [{xdim: 0, ydim: 1}, {xdim: 2, ydim: -1}];
const meow = svg.selectAll('.meow').data(data1, (d) => d.xdim);
meow.enter()
   ...//more code here
```

Above code will miss the change in xdim=1 from 3 to 1. For that reason, in the previous example, I meant to ask whether the data associated with xdim will never change.

Reply:
Usually TopoData changes across runs, but not across steps. The code works when I have two different runs with different topology. I'm not an expert at D3 yet, what is the suggested modification here?

Reply:
Sorry for being ambiguous. https://d3indepth.com/enterexit/
You can merge with cards up here and make below attr operations work in updates too.

Reply:
Ah, done. Thanks for tutorial!